js/misc/paygManager.js

```diff
@@ -0,0 +1,410 @@
+            // Don't use the setters here to prevent emitting a -changed signal
+            // on startup, which is useless and confuses the screenshield when
+            // selecting the session mode to construct the right unlock dialog.
+            this._enabled = this._proxy.Enabled;
+            this._expiryTime = this._proxy.ExpiryTime;
+            this._rateLimitEndTime = this._proxy.RateLimitEndTime;
+            this._setCodeFormat(this._proxy.CodeFormat || '^[0-9]{8}$');
+
+            this._propertiesChangedId = this._proxy.connect('g-properties-changed', this._onPropertiesChanged.bind(this));
+            this._codeExpiredId = this._proxy.connectSignal('Expired', this._onCodeExpired.bind(this));
+
+            this._maybeNotifyUser();
+            this._updateExpirationReminders();
+        }
+
+        this._initialized = true;
+        this.emit('initialized');
+    }
+
+    _onPropertiesChanged(proxy, changedProps) {
+        let propsDict = changedProps.deep_unpack();
+        if (propsDict.hasOwnProperty('Enabled'))
+            this._setEnabled(this._proxy.Enabled);
+
+        if (propsDict.hasOwnProperty('ExpiryTime'))
+            this._setExpiryTime(this._proxy.ExpiryTime);
+
+        if (propsDict.hasOwnProperty('RateLimitEndTime'))
+            this._setRateLimitEndTime(this._proxy.RateLimitEndTime);
+
+        if (propsDict.hasOwnProperty('CodeFormat'))
+            this._setCodeFormat(this._proxy.CodeFormat, true);
+    }
+
+    _setEnabled(value) {
+        if (this._enabled === value)
+            return;
+
+        this._enabled = value;
+        this.emit('enabled-changed', this._enabled);
+    }
+
+    _setExpiryTime(value) {
+        if (this._expiryTime === value)
+            return;
+
+        this._expiryTime = value;
+        this._updateExpirationReminders();
+
+        this.emit('expiry-time-changed', this._expiryTime);
+    }
+
+    _setRateLimitEndTime(value) {
+        if (this._rateLimitEndTime === value)
+            return;
+
+        this._rateLimitEndTime = value;
+        this.emit('rate-limit-end-time-changed', this._rateLimitEndTime);
+    }
+
+    _setCodeFormat(value, notify = false) {
+        if (this._codeFormat === value)
+            return;
+
+        this._codeFormat = value;
+        try {
+            this._codeFormatRegex = new GLib.Regex(
+                this._codeFormat,
+                GLib.RegexCompileFlags.DOLLAR_ENDONLY,
+                GLib.RegexMatchFlags.PARTIAL);
+        } catch (e) {
+            logError(e, 'Error compiling CodeFormat regex: %s'.format(this._codeFormat));
+            this._codeFormatRegex = null;
+        }
+
+        if (notify)
+            this.emit('code-format-changed');
+    }
+
+    _onCodeExpired() {
+        this.emit('code-expired');
+    }
+
+    _timeRemainingSecs() {
+        if (!this._enabled)
+            return Number.MAX_SAFE_INTEGER;
+
+        return Math.max(0, this._expiryTime - (GLib.get_real_time() / GLib.USEC_PER_SEC));
```

Comment:
This seems wrong. `g_get_real_time()` is documented to be based on wall clock time, whereas `EpgProvider:expiry-time` is based on `CLOCK_BOOTTIME`. `CLOCK_BOOTTIME` is the one we want to use; I don't think the shell should use wall clock time at all.

Reply:
Now I see there's a later commit which changes it to use CLOCK_BOOTTIME. We could squash in these two:

- "paygManager: Understand ExpiryTime in terms of CLOCK_BOOTTIME"
- "payg: Use CLOCK_BOOTTIME for rate limiting"

Reply:
The second commit was squashed with ba5b0ded0 but yeah, it makes sense to squash them.